In [118]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from mcdrift import *

We consider a toy example in which $N=3$, $P_0 = \begin{pmatrix}
0.2 & 0.3 & 0.5\\
0.1 & 0.8 & 0.1\\
0.3 & 0.4 & 0.3
\end{pmatrix}$, $P_1 = \begin{pmatrix}
0.8 & 0.1 & 0.1\\
0.6 & 0.2 & 0.2\\
0.4 & 0.1 & 0.5
\end{pmatrix}$ and $\pi = (1/3, 1/3, 1/3)$.

We set $T = 10000$, $t^* = 2000$, $L = 1000$ (for NP-CDM), $K = 3$ and $W = 5$

In [119]:
N = 3 # number of states
p0 = np.array([[0.2, 0.3, 0.5], [0.1, 0.8, 0.1], [0.3, 0.4, 0.3]]) # initial transition matrix
p1 = np.array([[0.8, 0.1, 0.1], [0.6, 0.2, 0.2], [0.4, 0.1, 0.5]]) # modified transition matrix
T = 10000 # sequence length
tstar = 2000 # time when abrupt change occurs
L = 1000 # number of observations guaranteed to come from p0
K = 3 # detection threshold
W = 5 # subsequence length
pi = np.array([1/3, 1/3, 1/3]) # initial distribution

seed = 2023 # for reproducibility

We first generate the sequence $\mathcal{T} = \{s_1, \cdots, s_T\}$.

In [120]:
np.random.seed(seed)

seq_sim = simulate_mc(pi, p0, tstar) # generate observations before shift

init_vec = np.zeros(N)
init_vec[seq_sim[-1]] = 1

seq_sim_2 = simulate_mc(init_vec, p1, T - tstar + 1) # generate observations after shift

seq_comb = seq_sim + seq_sim_2[1:] # combine observations into a list

Apply algorithm 1 (P-CDM)

In [121]:
#Arguments: sequence, W, p0, p1, K
pcdm(seq_comb, W, p0, p1, K)

402

The algorithm returns 402, which means it detects the shift in subsequence 402. In other word, the model predict that the shift occurs in $s_k$, where $2011 \leq k \leq 2015$.

Apply algorithm 2 (NP-CDM)

In [122]:
# Argument: sequence, W, N, L, K):
npcdm(seq_comb, W, N, L, K)

210

The algorithm returns 210, which means it detects the shift in subsequence 210. In other word, the model predict that the shift occurs in $s_k$, where $1051 \leq k \leq 1055$.